File: src/test/java/io/mfedirko/contactme/ContactFormValidationTest.java
package io.mfedirko.contactme;

import io.mfedirko.infra.RecaptchaClient;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import org.assertj.core.api.Condition;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;

import java.util.Collection;
import java.util.Set;

import static io.mfedirko.contactme.fixture.ContactForms.aContactForm;
import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest
public class ContactFormValidationTest {
    @Autowired
    private Validator validator;

    @MockBean
    private RecaptchaClient recaptchaClient;

    @Test
    void valid() {
        mockValidRecaptchaResponse();

        Set<ConstraintViolation<ContactForm>> violations = validator.validate(
                aContactForm().build());

        assertThat(violations).isEmpty();
    }

    @Test
    void missingRecaptcha() {
        Set<ConstraintViolation<ContactForm>> violations = validator.validate(
                aContactForm().recaptcha(null).build());

        assertThat(violations).has(fieldError("recaptcha"));
    }

    @Test
    void invalidRecaptcha() {
        mockInvalidRecaptchaResponse();

        Set<ConstraintViolation<ContactForm>> violations = validator.validate(
                aContactForm().build());

        assertThat(violations).has(fieldError("recaptcha"));
    }

    @Test
    void invalidEmail() {
        mockValidRecaptchaResponse();

        Set<ConstraintViolation<ContactForm>> violations = validator.validate(
                aContactForm().email("noeml").build());

        assertThat(violations).has(fieldError("email"));
    }

    @Test
    void invalidEmptyForm() {
        Set<ConstraintViolation<ContactForm>> violations = validator.validate(
                aContactForm().email("").fullName("").messageBody("").recaptcha(null).build());

        assertThat(violations)
                .has(fieldError("email"))
                .has(fieldError("fullName"))
                .has(fieldError("messageBody"))
                .has(fieldError("recaptcha"));
    }

    private void mockValidRecaptchaResponse() {
        Mockito.when(recaptchaClient.isValidCaptcha(Mockito.anyString()))
                .thenReturn(true);
    }

    private void mockInvalidRecaptchaResponse() {
        Mockito.when(recaptchaClient.isValidCaptcha(Mockito.anyString()))
                .thenReturn(false);
    }

    private static Condition<Collection<? extends ConstraintViolation<?>>> fieldError(String field) {
        return new Condition<>(el -> el.stream().anyMatch(
                cv -> field.equals(cv.getPropertyPath().toString())),
                "Validation error on field: " + field);
    }
}

File: src/test/kotlin/io/mfedirko/contactme/ContactFormValidationTest.kt
package io.mfedirko.contactme

import io.mfedirko.common.infra.RecaptchaClient
import io.mfedirko.fixture.ContactForms.aContactForm
import jakarta.validation.ConstraintViolation
import jakarta.validation.Validator
import org.assertj.core.api.Assertions
import org.assertj.core.api.Condition
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean

@SpringBootTest
class ContactFormValidationTest {
    @Autowired
    private lateinit var validator: Validator

    @MockBean
    private lateinit var recaptchaClient: RecaptchaClient
    
    @Test
    fun valid() {
        mockValidRecaptchaResponse()
        val violations = validator.validate(
            aContactForm()
        )
        Assertions.assertThat(violations).isEmpty()
    }

    @Test
    fun missingRecaptcha() {
        val violations = validator.validate(aContactForm().apply { recaptcha = "" })
        Assertions.assertThat(violations).has(fieldError("recaptcha"))
    }

    @Test
    fun invalidRecaptcha() {
        mockInvalidRecaptchaResponse()
        val violations = validator.validate(
            aContactForm()
        )
        Assertions.assertThat(violations).has(fieldError("recaptcha"))
    }

    @Test
    fun invalidEmail() {
        mockValidRecaptchaResponse()
        val violations = validator.validate(
            aContactForm().apply { email = "noeml" }
        )
        Assertions.assertThat(violations).has(fieldError("email"))
    }

    @Test
    fun invalidEmptyForm() {
        val violations: Set<ConstraintViolation<ContactForm>> = validator.validate(
            aContactForm().apply { email = ""; fullName = ""; messageBody = ""; recaptcha = "" }
        )
        Assertions.assertThat(violations)
            .has(fieldErrors("email", "fullName", "messageBody", "recaptcha"))
    }

    private fun mockValidRecaptchaResponse() {
        Mockito.`when`(recaptchaClient.isValidCaptcha(Mockito.anyString()))
            .thenReturn(true)
    }

    private fun mockInvalidRecaptchaResponse() {
        Mockito.`when`(recaptchaClient.isValidCaptcha(Mockito.anyString()))
            .thenReturn(false)
    }

    companion object {
        private fun fieldErrors(vararg fields: String): Condition<in (MutableCollection<out ConstraintViolation<*>>)> {
            return Condition(
                { violations -> fields.all {
                        field -> violations.any { it.propertyPath.toString() == field }
                    }
                },
                "Validation error on fields: $fields"
            )
        }

        private fun fieldError(field: String): Condition<in (MutableCollection<out ConstraintViolation<*>>)> {
            return Condition(
                { violations -> violations.any {
                        it.propertyPath.toString() == field
                    }
                },
                "Validation error on field: $field"
            )
        }
    }
}